README.md
# access-controller

[![Latest Release](https://img.shields.io/github/v/release/authorizer-tech/access-controller)](https://github.com/authorizer-tech/access-controller/releases/latest)
[![Go Report Card](https://goreportcard.com/badge/github.com/authorizer-tech/access-controller)](https://goreportcard.com/report/github.com/authorizer-tech/access-controller)
[![Slack](https://img.shields.io/badge/slack-%23authorizer--tech-green)](https://authorizer-tech.slack.com)

An implementation of a distributed access-control server that is based on [Google Zanzibar](https://research.google/pubs/pub48190/) - "Google's Consistent, Global Authorization System".

An instance of an `access-controller` is similar to the `aclserver` implementation called out in the paper. A cluster of access-controllers implement the functional equivalent of the Zanzibar `aclserver` cluster.

# Getting Started
If you want to setup an instance of the Authorizer platform as a whole, browse the API References, or just brush up on the concepts and design of the platform, take a look at the [official platform documentation](https://authorizer-tech.github.io/docs/overview/introduction). If you're only interested in running the access-controller then continue on.

## Setup a Cluster
An access-controller server supports single node or multi-node (clustered) topologies. Instructions for running the server with these topologies are outlined below.

To gain the benefits of the distributed query model that the access-controller implements, it is recommend to run a large cluster. Doing so will help distribute query load across more nodes within the cluster. The underlying cluster membership list is based on Hashicorp's [`memberlist`](https://github.com/hashicorp/memberlist)

> a library that manages cluster membership and member failure detection using a gossip based protocol.

A cluster should be able to suport hundreds of nodes. If you find otherwise, please [submit an issue](https://github.com/authorizer-tech/access-controller/issues/new).

### Docker Compose
[`docker-compose.yml`](./docker/docker-compose.yml) provides an example of how to setup a multi-node cluster using Docker and is a great way to get started quickly.

```console
$ docker compose -f docker/docker-compose.yml up
```

### Kubernetes (Recommended)
Take a look at our [official Helm chart](https://authorizer-tech.github.io/helm-charts/access-controller).

### Pre-compiled Binaries
Download the [latest release](https://github.com/authorizer-tech/access-controller/releases/latest) and extract it.

#### Pre-requisites
To run an access-controller you must have a running CockroachDB database. Take a look at setting up [CockroachDB with Docker](https://www.cockroachlabs.com/docs/stable/start-a-local-cluster-in-docker-mac.html).

#### Single Node
```console
$ ./bin/access-controller
```

#### Multi-node
Start a multi-node cluster by starting multiple independent servers and use the `-join` flag
to join the node to an existing cluster.

```console
$ ./bin/access-controller -node-port 7946 -grpc-port 50052
$ ./bin/access-controller -node-port 7947 -grpc-port 50053 -join 127.0.0.1:7946
$ ./bin/access-controller -node-port 7948 -grpc-port 50054 -join 127.0.0.1:7947
```

## Next Steps...
Take a look at the examples of how to:
* [Add a Namespace Configuration](https://authorizer-tech.github.io/docs/getting-started/add-namespace-config)
* [Write a Relation Tuple](https://authorizer-tech.github.io/docs/getting-started/write-relation-tuple)
* [Check a Subject's Access](https://authorizer-tech.github.io/docs/getting-started/check-access)

Don't hesitate to browse the official [Documentation](https://authorizer-tech.github.io/docs/overview/introduction), [API Reference](https://authorizer-tech.github.io/docs/api-reference/overview) and [Examples](https://authorizer-tech.github.io/docs/overview/examples/examples-intro).

# Community
The access-controller is an open-source project and we value and welcome new contributors and members
of the community. Here are ways to get in touch with the community:

* Slack: [#authorizer-tech](https://authorizer-tech.slack.com)
* Issue Tracker: [GitHub Issues](https://github.com/authorizer-tech/access-controller/issues)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.5] - 2021-08-30
### Fixed
* Allow the alias relation `...` to be implicitly defined.

## [0.1.4] - 2021-08-18
### Fixed
* Bug with cross namespace config snapshots not being uniquely stored per namespace within the context.

## [0.1.3] - 2021-07-09
### Removed
* `google.protobuf.FieldMask` option in the `ListRelationTuples` RPC. We can re-introduce
  it later if needed.

## [0.1.2] - 2021-06-28
### Added
* gRPC Health Checking

## [0.1.1] - 2021-06-25
### Added
* RPC input request validation

## [0.1.0] - 2021-06-18
### Added
* Initial `WriteService`, `ReadService`, `CheckService`, and `ExpandService` implementations
* Namespace config API and continuous namespace config snapshot monitoring
* Gossip based clustering and consistent hashing with bounded load-balancing for `Check` RPCs
* gRPC or HTTP/JSON (or both) API interfaces
* Kubernetes Helm chart

[Unreleased]: https://github.com/authorizer-tech/access-controller/compare/v0.1.5...HEAD
[0.1.5]: https://github.com/authorizer-tech/access-controller/compare/v0.1.4...v0.1.5
[0.1.4]: https://github.com/authorizer-tech/access-controller/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/authorizer-tech/access-controller/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/authorizer-tech/access-controller/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/authorizer-tech/access-controller/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/authorizer-tech/access-controller/releases/tag/v0.1.0

cmd/access-controller/main.go
package main

import (
	"context"
	"database/sql"
	"flag"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/signal"
	"runtime"
	"syscall"
	"time"

	"github.com/cenkalti/backoff/v4"
	"github.com/golang-migrate/migrate/v4"
	"github.com/golang-migrate/migrate/v4/database/cockroachdb"
	_ "github.com/golang-migrate/migrate/v4/source/file"
	"github.com/google/uuid"
	gwruntime "github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
	_ "github.com/lib/pq"
	"github.com/spf13/viper"

	aclpb "github.com/authorizer-tech/access-controller/genprotos/authorizer/accesscontroller/v1alpha1"
	ac "github.com/authorizer-tech/access-controller/internal"
	"github.com/authorizer-tech/access-controller/internal/datastores"
	"github.com/authorizer-tech/access-controller/internal/namespace-manager/postgres"
	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc"
	"google.golang.org/grpc/health/grpc_health_v1"
	"google.golang.org/grpc/reflection"
)

var (
	version = "dev"
	commit  = "none"
	date    = "unknown"
)

var serverID = flag.String("id", uuid.New().String(), "A unique identifier for the server. Defaults to a new uuid.")
var nodePort = flag.Int("node-port", 7946, "The bind port for the cluster node")
var advertise = flag.String("advertise", "", "The address that this node advertises on within the cluster")
var grpcPort = flag.Int("grpc-port", 50052, "The bind port for the grpc server")
var httpPort = flag.Int("http-port", 8082, "The bind port for the grpc-gateway http server")
var join = flag.String("join", "", "A comma-separated list of 'host:port' addresses for nodes in the cluster to join to")
var configPath = flag.String("config", "./localconfig/config.yaml", "The path to the server config")
var migrations = flag.String("migrations", "./db/migrations", "The absolute path to the database migrations directory")

type config struct {
	GrpcGateway struct {
		Enabled bool
	}

	CockroachDB struct {
		Host     string
		Port     int
		Database string
	}
}

func main() {

	flag.Parse()

	viper.SetConfigFile(*configPath)
	viper.AutomaticEnv()

	if err := viper.ReadInConfig(); err != nil {
		log.Fatalf("Failed to load server config file: %v", err)
	}

	var cfg config
	if err := viper.Unmarshal(&cfg); err != nil {
		log.Fatalf("Failed to Unmarshal server config: %v", err)
	}

	pgUsername := viper.GetString("POSTGRES_USERNAME")
	pgPassword := viper.GetString("POSTGRES_PASSWORD")

	dbHost := cfg.CockroachDB.Host
	if dbHost == "" {
		dbHost = "localhost"
		log.Warn("The database host was not configured. Defaulting to 'localhost'")
	}

	dbPort := cfg.CockroachDB.Port
	if dbPort == 0 {
		dbPort = 26257
		log.Warn("The database port was not configured. Defaulting to '26257'")
	}

	dbName := cfg.CockroachDB.Database
	if dbName == "" {
		dbName = "postgres"
		log.Warn("The database name was not configured. Defaulting to 'postgres'")
	}

	dsn := fmt.Sprintf("postgresql://%s:%s@%s:%d/%s?sslmode=disable",
		pgUsername,
		pgPassword,
		dbHost,
		dbPort,
		dbName,
	)

	db, err := sql.Open("postgres", dsn)
	if err != nil {
		log.Fatalf("Failed to establish a connection to the database: %v", err)
	}

	backoffPolicy := backoff.NewExponentialBackOff()
	backoffPolicy.MaxElapsedTime = 15 * time.Second

	err = backoff.Retry(func() error {
		if err := db.Ping(); err != nil {
			log.Error("Failed to Ping the database. Retrying again soon...")
			return err
		}

		return nil
	}, backoffPolicy)
	if err != nil {
		log.Fatalf("Failed to connect to the database: %v", err)
	}

	driver, err := cockroachdb.WithInstance(db, &cockroachdb.Config{})
	if err != nil {
		log.Fatalf("Failed to create migrator database driver instance: %v", err)
	}

	migrator, err := migrate.NewWithDatabaseInstance(
		fmt.Sprintf("file://%s", *migrations),
		"cockroachdb", driver)
	if err != nil {
		log.Fatalf("Failed to initialize the database migrator: %v", err)
	}

	if err := migrator.Up(); err != nil {
		if err != migrate.ErrNoChange {
			log.Fatalf("Failed to migrate up to the latest database schema: %v", err)
		}
	}

	datastore := &datastores.SQLStore{
		DB: db,
	}

	m, err := postgres.NewNamespaceManager(db)
	if err != nil {
		log.Fatalf("Failed to initialize postgres NamespaceManager: %v", err)
	}

	log.Info("Starting access-controller")
	log.Infof("  Version: %s", version)
	log.Infof("  Date: %s", date)
	log.Infof("  Commit: %s", commit)
	log.Infof("  Go version: %s", runtime.Version())

	ctrlOpts := []ac.AccessControllerOption{
		ac.WithStore(datastore),
		ac.WithNamespaceManager(m),
		ac.WithNodeConfigs(ac.NodeConfigs{
			ServerID:   *serverID,
			Advertise:  *advertise,
			Join:       *join,
			NodePort:   *nodePort,
			ServerPort: *grpcPort,
		}),
	}
	controller, err := ac.NewAccessController(ctrlOpts...)
	if err != nil {
		log.Fatalf("Failed to initialize the access-controller: %v", err)
	}

	healthChecker := ac.NewHealthChecker(controller.HealthCheck)

	addr := fmt.Sprintf(":%d", *grpcPort)
	listener, err := net.Listen("tcp", addr)
	if err != nil {
		log.Fatalf("Failed to start the TCP listener on '%v': %v", addr, err)
	}

	grpcOpts := []grpc.ServerOption{}
	server := grpc.NewServer(grpcOpts...)
	aclpb.RegisterCheckServiceServer(server, controller)
	aclpb.RegisterWriteServiceServer(server, controller)
	aclpb.RegisterReadServiceServer(server, controller)
	aclpb.RegisterExpandServiceServer(server, controller)
	aclpb.RegisterNamespaceConfigServiceServer(server, controller)
	grpc_health_v1.RegisterHealthServer(server, healthChecker)

	go func() {
		reflection.Register(server)

		log.Infof("Starting grpc server at '%v'..", addr)

		if err := server.Serve(listener); err != nil {
			log.Fatalf("Failed to start the gRPC server: %v", err)
		}
	}()

	var gateway *http.Server
	if cfg.GrpcGateway.Enabled {
		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		// Register gRPC server endpoint
		// Note: Make sure the gRPC server is running properly and accessible
		mux := gwruntime.NewServeMux()

		opts := []grpc.DialOption{grpc.WithInsecure()}

		if err := aclpb.RegisterCheckServiceHandlerFromEndpoint(ctx, mux, addr, opts); err != nil {
			log.Fatalf("Failed to initialize grpc-gateway CheckService handler: %v", err)
		}

		if err := aclpb.RegisterWriteServiceHandlerFromEndpoint(ctx, mux, addr, opts); err != nil {
			log.Fatalf("Failed to initialize grpc-gateway WriteService handler: %v", err)
		}

		if err := aclpb.RegisterReadServiceHandlerFromEndpoint(ctx, mux, addr, opts); err != nil {
			log.Fatalf("Failed to initialize grpc-gateway ReadService handler: %v", err)
		}

		if err := aclpb.RegisterNamespaceConfigServiceHandlerFromEndpoint(ctx, mux, addr, opts); err != nil {
			log.Fatalf("Failed to initialize grpc-gateway NamespaceConfig handler: %v", err)
		}

		gateway = &http.Server{
			Addr:    fmt.Sprintf(":%d", *httpPort),
			Handler: mux,
		}

		go func() {
			log.Infof("Starting grpc-gateway server at '%v'..", gateway.Addr)

			// Start HTTP server (and proxy calls to gRPC server endpoint)
			if err := gateway.ListenAndServe(); err != http.ErrServerClosed {
				log.Fatalf("Failed to start grpc-gateway HTTP server: %v", err)
			}
		}()
	}

	exit := make(chan os.Signal, 1)
	signal.Notify(exit, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)

	<-exit

	log.Info("Shutting Down..")

	if cfg.GrpcGateway.Enabled {
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()

		if err := gateway.Shutdown(ctx); err != nil {
			log.Errorf("Failed to gracefully shutdown the grpc-gateway server: %v", err)
		}
	}

	server.Stop()
	if err := controller.Close(); err != nil {
		log.Errorf("Failed to gracefully close the access-controller: %v", err)
	}

	log.Info("Shutdown Complete. Goodbye 👋")
}
